### PR TITLE
Lost password

### PR DESCRIPTION
**Pull request by pahrohfit**
Sample endpoint to send a password reset email (production shouldn't be showing failures like 'unknown email/user')
```
         try:
             notify = guard.send_reset_email(email)
             ret = {'message': 'password reset email sent for {}'.format(email)}
             return json_response("complete", ret)

         except Exception as e:
             logging.error("Error hit resetting user: {}".format(e))
             ret = {'message': 'failed to reset user',
                    'errors': repr(e)}
             return json_response("error", ret)
```

Sample endpoint for resetting a users password
```
         req = request.get_json(force=True)
         logging.debug("Reset Info: {}".format(req))

         try:
             user = guard.validate_reset_token(req['token'])
             user.update(password=guard.hash_password(req['new_pass']))
             ret = {'message': 'password reset good for: {}'.format(user['email'])}
             return json_response("complete", ret)

         except Exception as e:
             logging.error("Error hit resetting user by token: {}".format(e))
             return json_response("error", "error using that token")
```

### CODE REVIEW

File: flask_praetorian/base.py [hunk 1]
```diff
@@ -693,6 +833,32 @@ def get_user_from_registration_token(self, token):
...
         return user
 
+    def validate_reset_token(self, token):
+        """
+        Validates a password reset request based on the reset token
+        that is supplied. Verifies that the token is a reset token
+        and that the user can be properly retrieved
+        """
+        data = self.extract_jwt_token(token, access_type=AccessType.reset)
+        flask.current_app.logger.debug("DATA: {}".format(data))
```
Comment (dusktreader) [inline flask_praetorian/base.py]: left behind debugging log line

File: flask_praetorian/constants.py [hunk 2]
```diff
@@ -32,10 +41,12 @@
 
 REFRESH_EXPIRATION_CLAIM = 'rf_exp'
 IS_REGISTRATION_TOKEN_CLAIM = 'is_ert'
+IS_RESET_TOKEN_CLAIM = 'is_ert'
```
Comment (dusktreader) [inline flask_praetorian/constants.py]: seems like the reset token claim key should be different than the registration token claim key

File: flask_praetorian/base.py [hunk 3]
```diff
@@ -693,6 +833,32 @@ def get_user_from_registration_token(self, token):
...
+        PraetorianError.require_condition(
+            user_id is not None,
+            "Could not fetch an id from the reset token",
+        )
+        user = self.user_class.identify(user_id)
+        PraetorianError.require_condition(
+            user is not None,
+            "Could not identify the user from the reset token",
+        )
+        # TODO: this should probably just be mandatory to prevent replay
```
Comment (dusktreader) [inline flask_praetorian/base.py]: I think having a short lifespan for the token can mitigate this risk. I don't want to add another requirement for the user store. If the user wants to add logic to make sure that a reset token is only used once, that should probably be handled in their app.

My concern is that adding more requirements to the user store makes flask-praetorian less adaptable/user-friendly, especially for an edge case that might not be needed

File: tests/test_base.py [hunk 4]
```diff
@@ -733,6 +735,81 @@ def test_pack_header_for_user(self, app, user_class):
             assert token_data['duder'] == 'brief'
             assert token_data['el_duderino'] == 'not brief'
 
+    def test_reset_email(self, app, user_class, db, tmpdir):
+        """
+        This test verifies email based registration functions as expected.
```
Comment (dusktreader) [inline tests/test_base.py]: s/registration/reset/ (copy-pasta)

**Review by dusktreader — CHANGES_REQUESTED** [review]
Thank you very much for this contribution. I had a few issues, but nothing too serious. I'll check back in a bit to see how it's going

File: flask_praetorian/constants.py (hunk 2 above)
Comment (pahrohfit) [reply flask_praetorian/constants.py]: Good catch ... fixed

File: flask_praetorian/base.py (hunk 1 above)
Comment (pahrohfit) [reply flask_praetorian/base.py]: Yeap, cleaned.

File: tests/test_base.py (hunk 4 above)
Comment (pahrohfit) [reply tests/test_base.py]: yeah, i'm extra lazy in vim at times. cleaned.

File: flask_praetorian/base.py (hunk 3 above)
Comment (pahrohfit) [reply flask_praetorian/base.py]: I agree ... security hat wants to slap me, but for sanity sake with the new feature, best to error on the side of caution.  removed usages/tests.

**Comment by pahrohfit** [comment]
Think I addressed them all ... and added some missing tests around expiration time.

File: tests/test_base.py (hunk 4 above)
Comment (dusktreader) [reply tests/test_base.py]: ah, a fellow vim user. huzzah!

**Comment by dusktreader** [comment]
Looks like the build is passing in Travis, but the github connection is broken for some reason. Approving and merging

**Review by dusktreader — APPROVED** [review]
